### ISSUE TITLE
depthwise convolution to speed up

### ISSUE THREAD

**Issue by kaishijeng** [Robert-JunWang/Pelee#5]
Does pelee net use depthwise comvolution, like mobilenet?
If not, can it be modified to use it to speed up?

Thanks

**Comment by Robert-JunWang** [comment]
No. PeleeNet is built with conventional convolution. Depthwise convolution can reduce the number of multi-add. But for the real speed, it depends on the devices and the framework you used.  For example, both MobileNet+SSD and Pelee are much faster than TinyYOLO on CPU and iPhone6s, but TinyYOLO is slightly faster than Pelee on GPU and iPhone8. Although the number of multi-add of TinyYOLO is about 3 times larger than Pelee.

**Comment by Robert-JunWang** [comment]
Most of the work in this paper was completed 8 months ago. At that time, except Tensorflow, other frameworks had bad support for Depthwise Separatable Convolution. However, the situation has changed a lot now. The performance of grouped convolution has improved greatly on cuDNNv7.  Apple's CoreML also supports grouped convolution very well. 

It is a good time to try the Depthwise Separatable Convolution version now. I am more interested in improving the accuracy by increasing the number of channels with Depthwise Separatable Convolution. Both MobileNet and PeleeNet can perform the image classification on iphone6s, a phone released three years ago, in less than 50ms. The speed is good enough for many device-side applications.

**Comment by kaishijeng** [comment]
Pelee detection on my RK3399, (2 Core- arm A72 at 1.8Ghz) takes 400ms for a frame.
So performance is not sufficient for my application (<100ms)
Thanks,

**Comment by Robert-JunWang** [comment]
What framework do you use? If the framework doesn't support automatic
merging BN layer with Conv layer, you need to it by yourself. It can save
over 50% inference time. If you train the model from scratch, you can try a
wide and shallow network. For example, for the last two dense block, using
half of the number of the dense layer with a doubled growth rate.

**Comment by kaishijeng** [comment]
I use caffe framework and your pretrained model.
How do I check BN/Conv layer is merged or not?

Thanks,

**Comment by foralliance** [comment]
@Robert-JunWang 　HI

According to the model you provided, in `trian/test.prototxt`, **the BN/Scale layer** still exists alone. does this mean that conv layer and BN layer are not merged? If so, why is the inference time so fast?

The so-called "automatic merging BN layer with Conv layer", does it mean that we need to modify the underlying C++ code?

**Comment by Robert-JunWang** [comment]
The models I provided are not merged. You can merge it by hand or other tools. I can add the merged model and script I used next week.

**Comment by foralliance** [comment]
hope for your update

**Comment by ujsyehao** [comment]
@Robert-JunWang Hi, I have two problems:

- Have you uploaded merged model? or how to merge it?

- In your paper, it says > we use a shallow and wide network structure, Is there any theoretical support for this?

**Comment by Robert-JunWang** [comment]
Do you mean MobileNetV1+SSDlite is 10 times faster than MobileNetV1+SSD on CPU?  Do you mind to offer more detail information on how you evaluate the speed and how to get that result?  In my understanding, the computation cost of the SSD algorithm is mostly consumed by the backbone network. The real speed difference between SSDlite and the original SSD should not be that big.

**Comment by xonobo** [comment]
As long as I understand the given merged models do not contain any batch normalization. Just convolution and ReLU, right?

**Comment by lqs19881030** [comment]
> mute

@Robert-JunWang  what is the meaning of using half of the number of the dense layer with a doubled growth rate.can you show it and the map can fall much ?Thank you
